packages/gafl-webapp-service/src/pages/licence-details/licence-start-time/route.js

```diff
@@ -6,33 +6,55 @@ import moment from 'moment-timezone'
 import { cacheDateFormat } from '../../../processors/date-and-time-display.js'
 import { nextPage } from '../../../routes/next-page.js'
 
-const minHour = permission =>
-  moment(permission.licenceStartDate, cacheDateFormat)
+const getMinHour = permission => {
+  const now = moment().tz(SERVICE_LOCAL_TIME)
+  const permissionStartsToday = moment(permission.licenceStartDate, cacheDateFormat)
     .tz(SERVICE_LOCAL_TIME)
-    .isSame(moment().tz(SERVICE_LOCAL_TIME), 'day')
-    ? moment()
-      .tz(SERVICE_LOCAL_TIME)
-      .add(1, 'hour')
-      .add(30, 'minute')
-      .startOf('hour')
-      .hour()
-    : 0
-
-const hours = Array(24)
+    .isSame(now, 'day')
+  if (permissionStartsToday) {
+    const cantStartUntilTomorrow = moment(now).add(90, 'minute').isAfter(now, 'day')
+    if (cantStartUntilTomorrow) {
+      return 24
+    }
+    return now.add(90, 'minute').startOf('hour').hour()
+  }
+  return 0
+}
+
+const hours = Array(25)
   .fill(0)
   .map((e, idx) => idx.toString())
 
-// Not worth validating dynamically - can only be curl'd to the users disadvantage
-const validator = Joi.object({
-  'licence-start-time': Joi.string()
-    .valid(...hours)
-    .required()
-}).options({ abortEarly: false, allowUnknown: true })
+const validator = (payload, options) => {
+  const { permission } = options.context.app.request
+  const minHour = getMinHour(permission)
+
+  Joi.assert(
+    payload,
+    Joi.object({
+      'licence-start-time': Joi.string()
+        .valid(...hours.filter(h => h >= minHour))
+        .required()
+    }).options({ abortEarly: false, allowUnknown: true })
+  )
+}
 
 const getData = async request => {
   const permission = await request.cache().helpers.transaction.getCurrentPermission()
   const startDateStr = moment(permission.licenceStartDate, cacheDateFormat).format('dddd, Do MMMM, YYYY')
-  return { startDateStr, minHour: minHour(permission) }
+
+  return { startDateStr, minHour: getMinHour(permission) }
+}
+
+const route = pageRoute(LICENCE_START_TIME.page, LICENCE_START_TIME.uri, validator, nextPage, getData)
+route.find(r => r.method === 'POST').options.ext = {
+  onPostAuth: {
+    method: async (request, reply) => {
+      const permission = await request.cache().helpers.transaction.getCurrentPermission()
+      request.app.permission = permission
+      return reply.continue
```

Comment:
How come this is like this? Should this be in the update-transaction.js?

Reply:
No, because it's not actually modifying the permission, just making it available on the app object. In the validator, this is available whereas the `cache()` function isn't. The validator needs access to the permission as it needs the licence start date to compare to the current date.